micronota/parsers/gff.py

```diff
@@ -0,0 +1,238 @@
+
++------+------+---------------------------------------------------------------+
+|Reader|Writer|                          Object Class                         |
++======+======+===============================================================+
+|Yes   |No    |:mod:`skbio.metadata.IntervalMetadata` objects                 |
++------+------+---------------------------------------------------------------+
+|Yes   |No    |generator of :mod:`skbio.metadata.IntervalMetadata` objects    |
++------+------+---------------------------------------------------------------+
+
+
+Reference
+---------
+.. [#] http://gmod.org/wiki/GFF3
+'''
+
+from skbio.io import create_format, FileFormatError
+from skbio.metadata import IntervalMetadata, Feature
+from skbio.io.format._base import (
+    _line_generator, _get_nth_sequence, _too_many_blanks)
+
+
+class GFFFormatError(FileFormatError):
+    pass
+
+gff = create_format('gff')
+
+# Annotation headers
+_ANNOTATION_HEADERS = [
+          'SEQID',
+          'SOURCE',
+          'TYPE',
+          'SCORE',
+          'STRAND',
+          'PHASE',
+          'ATTR'
+          ]
+
+# Attribute headers
+_ATTR_HEADERS = [
+        'ID',
+        'Name',
+        'Alias',
+        'Parent',
+        'Target',
+        'Gap',
+        'Derives_from',
+        'Note',
+        'Dbxref',
+        'Ontology_term'
+        ]
+
+
+@gff.sniffer()
+def _gff_sniffer(fh):
+    # check the 1st real line is a valid ID line
+    if _too_many_blanks(fh, 5):
+        return False, {}
+
+    try:
+        line = next(_line_generator(fh, skip_blanks=True, strip=False))
+    except StopIteration:
+        return False, {}
+
+    try:
+        assert line.startswith('##gff-version 3')
+    except GFFFormatError:
+        return False, {}
+    return True, {}
+
+
+def _is_float(input):
+    try:
+        float(input)
+    except ValueError:
+        return False
+    return True
+
+
+def _parse_attr(s):
+    _field = s.split(';')
+    _attr = ()
+    for f in _field:
+        try:
+            _type, _val = f.split('=')
+        except ValueError:
+            _type, _val = 0, 0
+
+        _attr = _attr + (_type, _val)
+    return _attr
+
+
+def _parse_required(s):
+    if s.isdigit():
+        return int(s)
+    elif _is_float(s):
+        return float(s)
+    else:
+        return s
+
+
+def IntervalMetadata_construct(_attr, _interval):
+    im = IntervalMetadata()
+    im.add(Feature(**_attr), _interval)
+    return im
+
+
+def _construct(record, constructor=None, **kwargs):
+    attr, meta = record
+    if constructor is None:
+        constructor = IntervalMetadata_construct
+    return constructor(attr, meta)
+
+
+@gff.reader(None)
+def _gff_to_generator(fh, constructor=None, **kwargs):
+    for record in _parse_records(fh):
+        yield _construct(record, constructor, **kwargs)
+
+
+@gff.reader(IntervalMetadata)
+def _gff_to_metadata(fh, seq_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_records(fh), seq_num)
```

Comment:
do this in the import: `from skbio.xxx import _get_nth_sequence as _get_nth_record` to make the semantics clear